### PR TITLE
feat: log send payment simulation results to csv

### PR DESCRIPTION
**Pull request by okjodom**

### CODE REVIEW

**Comment by carlaKC** [comment]
Ran locally - succeeds for one row then gets unhappy.
```
ERROR [sim_lib] Error while reporting payment results: CsvError(Error(UnequalLengths { pos: None, expected_len: 8, len: 7 }))
```

File: 
```
source,destination,hash,amount_msat,dispatch_time,settled,htlc_count,failure_reason
0257956239efc55dd6be91eff40c47749314ccf79cb15f79e30ca12f8622b6de9e,02d804ad31429c8cc29e20ec43b4129553eb97623801e534ab5a66cdcd2149dbed,5e4e03a7c73d41f671b9231c17956920d7a0178ee80d1ed2f66d0f82118d52c2,100,1692337178455,true,0%
```

Payment:
```
Resolved payment received: Some((DispatchedPayment { source: PublicKey(9edeb622862fa10ce3795fb19cf7cc149374470cf4ef91bed65dc5ef39629557e8e67589509f07f628832580b4265dedd6a4063d6dcb0f71460276581d51c5d9), destination: PublicKey(eddb4921cdcd665aab34e501386297eb539512b443ec209ec28c9c4231ad04d85ac7469a940a609f4a331bb7a3430ac020b7ab0c813f4caf799565201b938efc), hash: PaymentHash([94, 78, 3, 167, 199, 61, 65, 246, 113, 185, 35, 28, 23, 149, 105, 32, 215, 160, 23, 142, 232, 13, 30, 210, 246, 109, 15, 130, 17, 141, 82, 194]), amount_msat: 100, dispatch_time: SystemTime { tv_sec: 1692337178, tv_nsec: 455801000 } }, PaymentResult { settled: true, htlc_count: 0, failure_reason: Some(RouteNotFound) }))
```

**Comment by okjodom** [comment]
> Ran locally - succeeds for one row then gets unhappy.
> 
> ```
> ERROR [sim_lib] Error while reporting payment results: CsvError(Error(UnequalLengths { pos: None, expected_len: 8, len: 7 }))
> ```
> 
> File:
> 
> ```
> source,destination,hash,amount_msat,dispatch_time,settled,htlc_count,failure_reason
> 0257956239efc55dd6be91eff40c47749314ccf79cb15f79e30ca12f8622b6de9e,02d804ad31429c8cc29e20ec43b4129553eb97623801e534ab5a66cdcd2149dbed,5e4e03a7c73d41f671b9231c17956920d7a0178ee80d1ed2f66d0f82118d52c2,100,1692337178455,true,0%
> ```
> 
> Payment:
> 
> ```
> Resolved payment received: Some((DispatchedPayment { source: PublicKey(9edeb622862fa10ce3795fb19cf7cc149374470cf4ef91bed65dc5ef39629557e8e67589509f07f628832580b4265dedd6a4063d6dcb0f71460276581d51c5d9), destination: PublicKey(eddb4921cdcd665aab34e501386297eb539512b443ec209ec28c9c4231ad04d85ac7469a940a609f4a331bb7a3430ac020b7ab0c813f4caf799565201b938efc), hash: PaymentHash([94, 78, 3, 167, 199, 61, 65, 246, 113, 185, 35, 28, 23, 149, 105, 32, 215, 160, 23, 142, 232, 13, 30, 210, 246, 109, 15, 130, 17, 141, 82, 194]), amount_msat: 100, dispatch_time: SystemTime { tv_sec: 1692337178, tv_nsec: 455801000 } }, PaymentResult { settled: true, htlc_count: 0, failure_reason: Some(RouteNotFound) }))
> ```
Needs 
- [ ] serde serialization of error reason
- [ ] continous flush of csv writer buffer to file (enables tail of csv?)
- [ ] guarantee write to file before simulation exit on conditions like SIGKILL

File: sim-lib/src/lib.rs [hunk 1]
```diff
@@ -220,6 +254,7 @@ impl Simulation {
             if let Err(e) = res {
                 log::error!("Task exited with error: {e}");
                 success = false;
+                shutdown_trigger.trigger();
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:257]: Why do we need to have the trigger here? If one of the threads exits (which means we're done awaiting) then we should have exited?

File: sim-lib/src/lib.rs [hunk 2]
```diff
@@ -414,18 +453,58 @@ async fn produce_events(
 }
 
 async fn consume_simulation_results(
-    mut receiver: Receiver<(DispatchedPayment, PaymentResult)>,
-    shutdown: triggered::Trigger,
+    receiver: Receiver<(DispatchedPayment, PaymentResult)>,
+    _shutdown: triggered::Trigger,
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:457]: Can remove shutdown trigger completely if we're not using it?

File: sim-lib/src/lib.rs [hunk 3]
```diff
@@ -414,18 +453,58 @@ async fn produce_events(
...
+        Ok(_) => {
+            log::debug!("Simulation results ok but exiting!!!");
+        }
+        Err(e) => log::error!("Error while reporting payment results: {:?}", e),
     }
+}
 
-    log::debug!("Simulation results consumer exiting");
-    shutdown.trigger();
+async fn log_payment_results(
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:470]: nit: write or flush payment_results? Since there's a logger coming in in #37.

File: sim-lib/src/lib.rs [hunk 4]
```diff
@@ -414,18 +453,58 @@ async fn produce_events(
...
+                        log::info!("Resolved payment received: {:?}", resolved_payment.clone());
+                        writer.serialize(resolved_payment).map_err(|e| {
+                            let _ = writer.flush();
+                            SimulationError::CsvError(e)
+                        })?;
+                        log::debug!("Serialized payment result");
+                        continue;
+                    },
+                    None => {
+                        log::debug!("Received none outcome");
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:501]: I think we can remove this + the serialized debug logs once we're happy it's all werkin!

**Review by carlaKC — COMMENTED** [review]
Just some cleanup - looking good!

File: sim-lib/src/lib.rs (hunk 1 above)
Comment (carlaKC) [reply sim-lib/src/lib.rs:257]: Tested without this in here and still works ✅

**Review by carlaKC — APPROVED** [review]
tACK, going to merge so that main has a full MVP. We can follow up with: 
* smarter batching
* write failure result

Also seems like `settled` is always true - I think we can update to just have a single `payment_error` which is 0 if we succeeded and an number representing an error code otherwise.

**Pull request merged**